### PR TITLE
Revisiting TestGrid Job statuses

### PR DESCRIPTION
**Pull request by pasindujw**
**Purpose**
Revisit the statuses in the TestGrid and do necessary modifications to make them consistent.

**Approach**
Each test-plan will maintain a status and a phase. 
From the starting of the test-grid job, each test-plan will go through following phases;
1. PREPARATION - Phase where the test-plan yaml is being prepared
2. INFRA_PHASE - Phase where the infrastructure of the test-plan is being created.
3. DEPLOY_PHASE - Phase where the deployment is being setup.
4. TEST_PHASE - Phase where the tests are being executed.

each of the phase above will have stages as _STARTED, SUCCEEDED, ERROR_
(e.g: `PREPARATION_PHASE_STARTED`, `TEST_PHASE_ERROR`)

Along with that, each test-plan will maintain a status (regardless of phases).
1. `RUNNING` - Test-plan is being running
2. `SUCCESS` - Test-plan has been executed successfully and all the tests have been passed.
3. `FAIL`         - Test-plan has been executed successfully and one or more tests have been failed.
4. `ERROR`    - Test-plan has not been executed successfully (Error in one of the phases above)

The log is implemented as;
At a given time,

- If the test-plan's status is one of { `SUCCESS`, `FAIL`} : That means all the phases have executed successfully and the current phase should be mentioned as `TEST_PHASE_SUCCEEDED` (which is the last phase of the flow)

- If the test-plan's status is `RUNNING`: test-plan is executing and the phase will be pointing to the currently running phase.

- If the test-plan's status is `ERROR`: test-plan has faced with some errors and the phase will be one of { `PREPARATION_ERROR`, `INFRA_PHASE_ERROR`, `DEPLOY_PHASE_ERROR`, `TEST_PHASE_ERROR` } which will imply the erroneous phase.

**Important note:**
- By telling `TEST_PHASE_SUCCEEDED` doesn't mean the tests are "passing". It only says that all the tests have been executed properly. The results of the tests will be shown only by the status of the Test-Plan { `SUCCESS` if all are passing, `FAIL` if one or more failing }

- Even if all the identified tests are passing/failing, the status of the test-plan could be `ERROR` if at least one of the phases have faced errors.

**Security checks**
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

**Required db change**

> alter table test_plan add column phase varchar(100);

### CODE REVIEW

File: core/src/main/java/org/wso2/testgrid/core/TestPlanExecutor.java [hunk 1]
```diff
@@ -132,69 +134,89 @@ public boolean execute(TestPlan testPlan, InfrastructureConfig infrastructureCon
...
             GrafanaDashboardHandler dashboardSetup = new GrafanaDashboardHandler(testPlan.getId());
             dashboardSetup.initDashboard();
         }
 
         // Create and set deployment.
-        DeploymentCreationResult deploymentCreationResult = createDeployment(testPlan,
-                infrastructureProvisionResult);
+        if (testPlan.getPhase().equals(TestPlanPhase.INFRA_PHASE_SUCCEEDED) &&
+                testPlan.getStatus().equals(TestPlanStatus.RUNNING)) {
+            persistTestPlanPhase(testPlan, TestPlanPhase.DEPLOY_PHASE_STARTED);
```
Comment (kasunbg) [inline core/src/main/java/org/wso2/testgrid/core/TestPlanExecutor.java]: If possible, I'd like to keep the #execute method as clean and short as possible. Technically, it should only have four code lines - infra create, deploy create, test run, infra destroy. We won't be able to go to that level, but let's see what we can do.

File: core/src/main/java/org/wso2/testgrid/core/TestPlanExecutor.java [hunk 2]
```diff
@@ -132,69 +134,89 @@ public boolean execute(TestPlan testPlan, InfrastructureConfig infrastructureCon
...
+            if (!deploymentCreationResult.isSuccess()) {
+                testPlan.setStatus(TestPlanStatus.ERROR);
+                testPlan.setPhase(TestPlanPhase.DEPLOY_PHASE_ERROR);
+                for (ScenarioConfig scenarioConfig : testPlan.getScenarioConfigs()) {
+                    scenarioConfig.setStatus(Status.DID_NOT_RUN);
+                }
+                testPlanUOW.persistTestPlan(testPlan);
+                logger.error(StringUtil.concatStrings(
+                        "Error occurred while performing deployment for test plan", testPlan.getId(),
+                        "Releasing infrastructure..."));
```
Comment (kasunbg) [inline core/src/main/java/org/wso2/testgrid/core/TestPlanExecutor.java]: Missing spaces.

File: core/src/main/java/org/wso2/testgrid/core/TestPlanExecutor.java (hunk 1 above)
Comment (pasindujw) [reply core/src/main/java/org/wso2/testgrid/core/TestPlanExecutor.java]: Refactored as much as possible to accomplish that :)

Comment (kasunbg) [reply core/src/main/java/org/wso2/testgrid/core/TestPlanExecutor.java]: Nice.. That's a good design!

**Pull request merged**